### DISCUSSION TITLE
Best way to define new reusable schemas?

### DISCUSSION THREAD

**Discussion by prescience-data** [colinhacks/zod#941] [Q&A] [answered]
Often I have an issue where I need to make a schema that is used frequently within an `z.object` definition.

An example is parsing environment variables where the initial value is always a string, but should be a number at the end of parsing. 

The issue is that I want to provide a default value so (to the best of my knowledge) I need to provide the `.default()` call _before_ the `.transform()` call, which means that just extending a `z.string().transform()` call won't work(?).

My current solution is below but it seems kind of janky, and has the disadvantage of _requiring_ a default to be set (eg the only interface available in this case is `numberString().default(100)`) and I can't see simple way of providing simply `numberString()` without redefining the schema within the export.

### Example
```typescript
export const numberString = () => {
  return {
    default: (defaultValue) =>
      z
        .union([z.string(), z.number()])
        .default(`${defaultValue}`)
        .transform((value) => parseInt(`${value}`))
  }
}

```

Is there some sort of "constructor" for building new schemas that provide all the usual behaviours that I've missed? Or possibly a better way of defining the above to make it more flexible?

**Comment by JacobWeisenburger** [comment, marked as answer]
I think you could use [preprocess](https://github.com/colinhacks/zod#preprocess).

Let me know if you have any questions.

```ts
const stringToNumber = () => z.preprocess(
    input => {
        if ( typeof input == 'number' ) return input
        if ( typeof input == 'string' ) {
            if ( input.includes( '.' ) ) // if you want to allow decimals
                return parseFloat( input )
            else
                return parseInt( input )
        }
    },
    z.number()
)
```

```ts
const portSchema = stringToNumber().default( 3000 )
type PortSchemaInput = z.input<typeof portSchema>
// type PortSchemaInput = number | undefined
type PortSchemaOutput = z.output<typeof portSchema>
// PortSchemaOutput = number

console.log( portSchema.parse( undefined ) ) // 3000
console.log( portSchema.parse( '4200' ) )    // 4200
console.log( portSchema.parse( 4200 ) )      // 4200

console.log( portSchema.safeParse( '' ).success )   // false
console.log( portSchema.safeParse( null ).success ) // false
```

```ts
const ageSchema = stringToNumber()
type AgeSchema = z.infer<typeof ageSchema>
// type AgeSchema = number

console.log( ageSchema.parse( '42' ) ) // 42   
console.log( ageSchema.parse( 42 ) )   // 42   
console.log( ageSchema.parse( 0 ) )    // 0

console.log( ageSchema.safeParse( undefined ).success ) // false
console.log( ageSchema.safeParse( '' ).success )        // false
console.log( ageSchema.safeParse( null ).success )      // false
```
